[FEAT] 멘토링 페이지 평점 높은 순 정렬

kimdevv opened the pull request:
## Issue Number
closed #779 

## As-Is
* 정렬 기준이 최신순, 예약 많은 순밖에 없었습니다.

## To-Be
* 멘토링 별 평균 별점이 높은 순으로 정렬해서 조회하는 기능이 추가되었습니다.
* `MentoringStatistics` 테이블에 `averageRating` 필드가 추가되었습니다.
  - `ratingSum`과 `reviewCount`로 매 조회 요청마다 평균 별점을 구해서 정렬하여 반환할 수 있었습니다. 단순 연산이기 때문에 연산 속도는 크게 느려지지 않습니다.
  - 그러나 해당 방식으로는 평균 별점에 인덱스를 생성할 수 없어서 조회 성능이 느려질 것이라고 생각했습니다.
  - 그래서 리뷰가 생성/삭제될 때마다 각 `mentoringStatistics` 레코드의 `reviewCount`와 `ratingSum`을 업데이트해주고, 해당 필드들을 통해 `averageRating`을 계산하여 갱신합니다(`ratingSum` / `reviewCount`).

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?

kimdevv committed:
feat: 평균 평점 기준 정렬 기능 추가

- SortKey에 AVERAGE_RATING 값 추가
- 평균 평점을 구해 정렬 조건에 추가
- 평균 평점 기준의 커서 코드 생성 메서드 추가
- 주석 문구 수정 및 보완

kimdevv committed:
test: 멘토링 별점 평균 정렬 및 카테고리 필터링 테스트 추가

- 별점 평균 순으로 멘토링 목록 페이지 조회 테스트 추가
- 카테고리 필터링 조합과 별점 평균 정렬 기능 테스트 추가
- 테스트 데이터 및 관련 엔티티 초기화 로직 보완

ppparkta commented on backend/fittoring/src/test/java/fittoring/integration/MentoringIntegrationTest.java:
p3: 연관관계로 묶인 엔티티가 많아서 테스트 두 개가 엄청 기네요! 고생하셨어요.

ppparkta reviewed (approved):
처음엔 읽기 어려웠던 코드가 익숙해지니 읽혀서 신기하네요! 구조 유지한 상태로 정렬 조건만 추가돼서 많이 깔끔한 것 같아요. 

코치님들이 말씀해주신 것처럼 리뷰 평균이 5점인 경우가 많아질 것 같네요. 그런 경우를 대비해서 새로운 정렬 정책을 세울 수 있을 것 같아요. 통계 테이블에 추가적인 정보를 담으면 정책 세우기가 더 수월할 것 같습니다. 이 부분은 2차 스프린트에서 함께 논의해보면 좋을 것 같아요 😄 

이상 없어보여서 바로 승인합니다. 고생하셨어요!

kimdevv committed:
feat: 멘토링 통계 평균 별점 저장 및 쿼리 최적화

- MentoringStatistics에 averageRating 필드 추가 및 관련 getter 생성
- 평균 별점 계산 로직을 쿼리에서 처리하도록 변경하여 최적화
- DB 마이그레이션 추가 (mentoring_statistics 테이블에 인덱스 및 average_rating 필드 추가)
- 기존 averageRating 계산 메서드 제거 및 관련 코드 업데이트
- 정렬 및 커서 코드 생성 로직에서 averageRating 필드를 직접 활용하도록 수정

kimdevv commented on backend/fittoring/src/main/java/fittoring/application/mentoring/repository/MentoringStatisticsRepository.java:
```diff
@@ -22,27 +22,32 @@ public interface MentoringStatisticsRepository extends ListCrudRepository<Mentor
-                ms.ratingSum = ms.ratingSum - :rating
-            WHERE ms.id = :mentoringId
-        """)
+    @Query(value = """
+            UPDATE mentoring_statistics ms
+            SET ms.review_count = ms.review_count - 1,
+                ms.rating_sum = ms.rating_sum - :rating,
+                ms.average_rating = COALESCE(
+                        ms.rating_sum / NULLIF(ms.review_count, 0),
+                        0.0
+               )
+            WHERE ms.mentoring_id = :mentoringId
+        """, nativeQuery = true)
```
리뷰가 삭제될 때 `reviewCount`를 -1 해주고 ratingSum도 그만큼 차감합니다. 그리고 그 이후 `ratingSum` / `reviewCount`를 통해 새로운 `averageRating`을 구합니다. 이 때 `reviewCount`가 0이 될 수 있어서, Divide By Zero를 방지하기 위해 `COALESCE()`, `NULLIF()` 함수와 함께 네이티브 쿼리를 적용했습니다.
해당 네이티브 쿼리는 `ratingSum`과 `reviewCount`을 나눠 `averageRating`을 구하되, `reviewCount`가 0이면 `ratingSum` / `NULL`을 만들어 `COALESCE()` 함수의 첫 번째 인자를 `NULL`로 바꿔서 `0.0`을 반환하도록 합니다.

쿼리가 복잡해지는 것을 막으려면 서비스 레이어에서 자바 코드로 작성할 수도 있는데, 그렇게 되면 SQL을 2번을 날려야 해서 그냥 이렇게 한 번의 SQL으로 처리하는 방식을 선택했습니다. 의견 잇으시면 남겨주세요~

ppparkta replied:
p3: 리뷰 삭제로 인해 0으로 나눌 때의 예외처리도 꼼꼼히 해주셨네요. 개인적으로 쿼리가 복잡하지 않은 것 같기도 하고 쿼리를 두번 날리는 방식이면 `averageRating`의 정합성을 매번 신경써줘야 할 것 같아서 지금 방식이 좋습니다.

dye0p replied:
쉽게 설명해주셔서 감사해요!

ppparkta commented on backend/fittoring/src/main/java/fittoring/domain/model/MentoringStatistics.java:
```diff
@@ -41,6 +41,10 @@ public class MentoringStatistics {
     @Column(nullable = false)
     private long ratingSum;
 
+    @Getter
+    @Column(nullable = false)
+    private double averageRating;
```
p3: 설명해주신대로 인덱스 등 성능 개선을 위해 평점을 함께 관리하게 되었네요. 좋은 방법 같습니다 👍

jye326 replied:
p3: 별점 평균에 인덱스도 걸 수 있고 좋네요. 👍

ppparkta reviewed (approved):
예외처리 꼼꼼히 잘해주셨네요. 고생하셨어요! 통계 테이블이 도입되면서 할 수 있는 것들이 많아질 것 같아요. 핏토링도 데이터 관리에 힘써봅시다 파잍팅!

jye326 commented on backend/fittoring/src/main/java/fittoring/domain/model/SortKey.java:
```diff
@@ -4,5 +4,6 @@ public enum SortKey {
 
     CREATED_AT,
     RESERVATION_COUNT,
+    AVERAGE_RATING,
```
벌써 세 번째 정렬기준이네요👍

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/business/repository/helper/MentoringPaginationHelper.java:
```diff
@@ -69,6 +69,12 @@ private BooleanExpression buildCursorCondition(SortKey sortKey, Cursor cursor) {
                 .or(mentoringStatistics.reservationCount.eq(cursorReservationCount)
                         .and(mentoring.id.loe(cursor.id())));
         }
+        if (sortKey == SortKey.AVERAGE_RATING && cursor != null) {
+            double cursorAverageCount = Double.longBitsToDouble(cursor.sortValue());
+            return mentoringStatistics.averageRating.lt(cursorAverageCount)
+                    .or(mentoringStatistics.averageRating.eq(cursorAverageCount)
+                        .and(mentoring.id.loe(cursor.id())));
+        }
```
p2: cursor가 null일 경우를 미리 return하고, 나머지는 아래의 `buildOrderSpecifiers`처럼 switch(sortKey) 문으로 처리하면 가독성을 더 높일 수 있을 것 같아요!

kimdevv replied:
그렇겠네요. 반영하겠습니다~

jye326 commented on backend/fittoring/src/main/java/fittoring/application/mentoring/repository/MentoringPaginationHelper.java:
```diff
@@ -129,13 +139,14 @@ public String generateNextCursorCode(SortKey sortKey, boolean hasNext, List<Tupl
 
     /**
-     * 정렬 기준이 created_at인 경우의 다음 커서를 문자열한다.
+     * 정렬 기준이 created_at인 경우의 다음 커서를 문자열화 한다.
```
👁️

jye326 commented on backend/fittoring/src/main/java/fittoring/application/mentoring/repository/MentoringStatisticsRepository.java:
```diff
@@ -22,27 +22,32 @@ public interface MentoringStatisticsRepository extends ListCrudRepository<Mentor
-                ms.ratingSum = ms.ratingSum - :rating
-            WHERE ms.id = :mentoringId
-        """)
+    @Query(value = """
+            UPDATE mentoring_statistics ms
+            SET ms.review_count = ms.review_count - 1,
+                ms.rating_sum = ms.rating_sum - :rating,
+                ms.average_rating = COALESCE(
+                        ms.rating_sum / NULLIF(ms.review_count, 0),
+                        0.0
+               )
+            WHERE ms.mentoring_id = :mentoringId
+        """, nativeQuery = true)
```
p3 : 복잡한 쿼리네요. QueryDSL을 적용해 볼 만한 것 같아요!

kimdevv replied:
추후에 다같이 이야기 해보고 적용해보면 좋을 것 같습니다!

jye326 reviewed (commented):
수고하셨습니다! 코멘트 남겼으니 확인해주시고 develop 컨플릭트 해결 후 다시 요청 주세요~

kimdevv committed:
Merge branch 'develop' into feature/779-order-most-rating

# Conflicts:
#	backend/fittoring/src/main/java/fittoring/application/mentoring/service/MentoringService.java

kimdevv committed:
refactor: buildCursorCondition 메서드 로직 리팩토링

- sortKey에 따라 switch-case문으로 분기 처리하도록 수정
- null 체크 로직 개선 및 가독성 향상
- 반복되는 cursor 처리 코드 중복 제거
- 코드 구조화로 유지보수성 강화

dye0p commented on backend/fittoring/src/main/java/fittoring/application/mentoring/repository/MentoringPaginationHelper.java:
```diff
@@ -55,19 +55,30 @@ private BooleanExpression buildSoftDeleteCondition() {
-            long cursorReservationCount = cursor.sortValue();
-            return mentoringStatistics.reservationCount.lt(cursorReservationCount)
-                .or(mentoringStatistics.reservationCount.eq(cursorReservationCount)
+        switch (sortKey) {
+            case CREATED_AT -> {
+                LocalDateTime cursorDateTime = Instant.ofEpochSecond(cursor.sortValue())
+                    .atZone(ZoneId.of("Asia/Seoul"))
+                    .toLocalDateTime();
+                return mentoring.createdAt.lt(cursorDateTime)
+                    .or(mentoring.createdAt.eq(cursorDateTime)
+                        .and(mentoring.id.loe(cursor.id())));
+            }
+            case RESERVATION_COUNT -> {
+                long cursorReservationCount = cursor.sortValue();
+                return mentoringStatistics.reservationCount.lt(cursorReservationCount)
+                    .or(mentoringStatistics.reservationCount.eq(cursorReservationCount)
                         .and(mentoring.id.loe(cursor.id())));
+            }
+            case AVERAGE_RATING -> {
+                double cursorAverageCount = Double.longBitsToDouble(cursor.sortValue());
+                return mentoringStatistics.averageRating.lt(cursorAverageCount)
+                    .or(mentoringStatistics.averageRating.eq(cursorAverageCount)
+                        .and(mentoring.id.loe(cursor.id())));
+            }
```
정렬기준의 개수에 따라서 case가 늘어나고 있네요. 아직은 분기처리를 해도 괜찮을 것 같습니다! 추후에 정렬기준이 더 생긴다면 인터페이스 분리도 생각해보면 좋을 것 같아요~

dye0p commented on backend/fittoring/src/test/java/fittoring/integration/MentoringIntegrationTest.java:
```diff
@@ -1207,6 +1207,286 @@ void getMentoringSummaryPagesWithCategory2() {
             });
         }
 
+        @DisplayName("첫 멘토링 목록 페이지를 별점 평균 높은 순으로 조회하고, 반환된 커서를 사용하면 나머지 목록을 조회할 수 있다.")
```
테스트 꼼꼼하게 작성해주셔서 감사해요👍

dye0p commented on backend/fittoring/src/main/resources/db/migration/V202510021103__add_indexes_of_mentoring_statistics.sql:
```diff
@@ -0,0 +1,8 @@
+ALTER TABLE mentoring_statistics
+    ADD COLUMN average_rating DOUBLE NOT NULL DEFAULT 0;
+
+ALTER TABLE mentoring ADD INDEX idx_created_at (created_at);
+
+ALTER TABLE mentoring_statistics
+    ADD INDEX idx_reservation_count (reservation_count),
+    ADD INDEX idx_average_rating (average_rating);
```
인덱스 좋아요👍

dye0p reviewed (approved):
정렬 조건 추가해주시느라 수고 많으셨어요! 코드에 큰 문제 없어보여서 머지하겠습니다! 수고하셨습니다~
